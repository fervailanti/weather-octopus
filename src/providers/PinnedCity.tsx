import React, { FC, useReducer, createContext, ReactElement } from 'react'
import { PinnedCityAction, PinnedCityContextType, PinnedCityType } from '../helpers/types'

const initialState: PinnedCityType = {
  id: 6359304,
  coord: {
    lon: -3.68275,
    lat: 40.489349
  },
  isLoading: true
}

const initialPinnedCity: PinnedCityType = JSON.parse(
  localStorage.getItem('pinnedCity') || JSON.stringify(initialState)
)

export const PinnedCityContext = createContext<PinnedCityContextType>({
  pinnedCity: initialPinnedCity,
  setPinnedCity: () => {}
})

const PinnedCityReducer = (
  prevState: PinnedCityType = initialPinnedCity,
  action: PinnedCityAction
) => {
  const actions: any = {
    id: {
      ...prevState,
      id: action.id
    },
    coord: {
      ...prevState,
      coord: action.coord
    },
    loading: {
      ...prevState,
      isLoading: action.loading
    }
  }
  localStorage.setItem('pinnedCity', JSON.stringify(actions[action.type]))
  return actions[action.type]
}

export const PinnedCityProvider: FC = ({ children }): ReactElement => {
  const [pinnedCity, setPinnedCity] = useReducer(PinnedCityReducer, initialPinnedCity)
  const value = { pinnedCity, setPinnedCity }
  return <PinnedCityContext.Provider value={value}>{children}</PinnedCityContext.Provider>
}
