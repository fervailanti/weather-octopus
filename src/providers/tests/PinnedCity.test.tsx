import React from 'react'
import { PinnedCityProvider, pinnedCityReducer } from '../PinnedCity'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

describe('Pinned City Provider (state)', () => {
  it.each([
    // ---- TEST CASES: ----
    // CASE 1:
    [
      // Prev State:
      {
        id: 2345,
        coord: {
          lon: 55,
          lat: 40
        },
        isLoading: false
      },
      // Action:
      {
        type: 'id',
        id: 6543
      },
      // Expected State:
      {
        id: 6543,
        coord: {
          lon: 55,
          lat: 40
        },
        isLoading: false
      }
    ],
    // CASE 2:
    [
      // Prev State:
      {
        id: 2345,
        coord: {
          lon: 55,
          lat: 40
        },
        isLoading: false
      },
      // Action:
      {
        type: 'coord',
        coord: {
          lon: 20,
          lat: 75
        }
      },
      // Expected State:
      {
        id: 2345,
        coord: {
          lon: 20,
          lat: 75
        },
        isLoading: false
      }
    ],
    // CASE 3:
    [
      // Prev State:
      {
        id: 2345,
        coord: {
          lon: 55,
          lat: 40
        },
        isLoading: false
      },
      // Action:
      {
        type: 'loading',
        loading: true
      },
      // Expected State:
      {
        id: 2345,
        coord: {
          lon: 55,
          lat: 40
        },
        isLoading: true
      }
    ]
  ])('Reducer should handle state correctly', (prevState, action, expectedState) => {
    expect(pinnedCityReducer(prevState, action)).toStrictEqual(expectedState)
  })

  it('PinnedCityProvider Component should render correctly', () => {
    const wrapper = render(
      <PinnedCityProvider>
        <p>TEST</p>
      </PinnedCityProvider>
    )
    expect(wrapper.container).toMatchSnapshot()
  })
})
