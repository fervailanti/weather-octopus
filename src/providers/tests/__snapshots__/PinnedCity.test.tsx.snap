// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pinned City Provider (state) PinnedCityProvider Component should render correctly 1`] = `
<div>
  <p>
    TEST
  </p>
</div>
`;
